frontend/test/pytest/test_callback.py

```diff
@@ -586,5 +587,269 @@ def wrapper():
+
+    captured = capsys.readouterr()
+    assert str(arg) in captured.out.strip()
+
+
+@pytest.mark.parametrize("scale", [(0.1), (0.2), (0.3)])
+def test_active_grad_no_tape(scale):
+    """Test that pure callback can be differentiated no tape"""
+
+    @pure_callback
+    def identity(x) -> float:
+        return x
+
+    @identity.fwd
+    def fwd(x):
+        # Still needs to return a tuple.
+        return identity(x), None
+
+    @identity.bwd
+    def bwd(_res, cot):
+        return cot
+
+    @qml.qjit
+    @grad
+    def wrapper(x):
+        return scale * identity(x)
+
+    assert np.allclose(wrapper(42.0), scale)
+
+
+@pytest.mark.parametrize("scale", [(0.1), (0.2), (0.3)])
+def test_active_grad_tape(scale):
+    """Test that pure callback can be differentiated with a tape"""
+
+    @pure_callback
+    def identity(x) -> float:
+        return x
+
+    @identity.fwd
+    def fwd(x):
+        return identity(x), 1.0
+
+    @identity.bwd
+    def bwd(res, cot):
+        return cot * res
+
+    @qml.qjit
+    @grad
+    def wrapper(x):
+        return scale * identity(x)
+
+    assert np.allclose(wrapper(42.0), scale)
+
+
+@pytest.mark.parametrize("scale", [(0.1), (0.2), (0.3)])
+@pytest.mark.parametrize("space", [(2), (3), (4)])
+def test_active_grad_many_residuals(scale, space):
+    """Test that pure callback can be differentiated with many residuals"""
+
+    @pure_callback
+    def identity(x) -> float:
+        return x
+
+    @identity.fwd
+    def fwd(x):
+        tape = [1 / space] * space
+        return identity(x), tuple(tape)
+
+    @identity.bwd
+    def bwd(res, cot):
+        return cot * sum(res)
+
+    @qml.qjit
+    @grad
+    def wrapper(x):
+        return scale * identity(x)
+
+    assert np.allclose(wrapper(42.0), scale)
+
+
+@pytest.mark.parametrize("scale", [(0.1), (0.2), (0.3)])
+@pytest.mark.parametrize("space", [(2), (3), (4)])
+def test_active_jacobian_many_residuals(scale, space):
+    """Test that pure callback can be differentiated with many residuals"""
+
+    # This is a hack, just for the type
+    arg = jax.numpy.identity(2, dtype=float)
+
+    @pure_callback
+    def identity(x) -> arg:
+        return x
+
+    @identity.fwd
+    def fwd(x):
+        tape = [1 / space] * space
+        return identity(x), tuple(tape)
+
+    @identity.bwd
+    def bwd(res, cot):
+        return cot * sum(res)
+
+    @qml.qjit
+    @jacobian
+    def wrapper(x):
+        return scale * identity(x)
+
+    @jax.jit
+    @jax.jacobian
+    def wrapper_jax(x):
+        return scale * x
+
+    assert np.allclose(wrapper(arg), wrapper_jax(arg))
+
+
+@pytest.mark.parametrize("arg0", [(0.1), (0.2), (0.3)])
+@pytest.mark.parametrize("arg1", [(2.0), (3.0), (4.0)])
+def test_example_from_story(arg0, arg1):
+    """Just exactly the same function on the spec
+    modulo errors in the example
+    """
+
+    @pure_callback
+    def some_func(x, y) -> float:
+        return np.sin(x) * y
```

Comment:
Hey @erick-xanadu, apologies for not catching in the in the story, but it would be great to copy and paste this test with a slight modification, where the input is an array:

```python
@pure_callback
def some_func(x) -> float:
    return np.sin(x[0]) * x[1]

@some_func.fwd
def some_func_fwd(x):
    return some_func(x), (jnp.cos(x[0]), jnp.sin(x[0]), x[1])

@some_func.bwd
def some_func_bws(res, dy):
    cos_x0, sin_x0, y0 = res  # Gets residuals computed in f_fwd
    # since there is a single array parameter, we return
    # a VJP which is an array parameter of the same shape
    return jnp.array([cos_x * dy * y, sin_x * dy])

@qml.qjit
@grad
def cost(x):
    y = jnp.array([jnp.cos(x[0]), x[1]])
    return jnp.sin(some_func(y))
```

Let me know if by some chance this isn't working (no need to dig into it), and I'll track it on SC.

Reply:
Another one that could be good to add a test for is `some_func(x: Array, y: Array) -> Array`, but I'll need to craft an example of this so no worries on this one.

Reply:
@erick-xanadu I've written up all test cases I can think of here: https://gist.github.com/josh146/b65436b71221a12df9a1ad9413df2314

Reply:
I've added them all with minor differences. None of the differences seem to be significant (i.e., the error appears to be in the test not on the code). But it might be worthwhile to double check.

Reply:
I'll also add an issue to add a warning if the user uses a shape with dtype=float32. I am pretty sure we must only use float64.

Reply:
I added the one requested by David, and I tried also differentiating A @ B and comparing it to JAX. The second one is currently producing wrong results. This could be related to the same issue we've been seeing in the past regarding not passing metadata to Enzyme but I am not 100% sure yet.

Reply:
https://github.com/PennyLaneAI/catalyst/pull/834/commits/c1916125a8972c374de161f554089e394996789c

Reply:
Thanks @erick-xanadu! Could you open an issue for the failing test example if you haven't already (too many notifications atm 😅)